infer/component.go

```diff
@@ -30,40 +30,72 @@ import (
 	"github.com/pulumi/pulumi-go-provider/middleware/schema"
 )
 
+// ComponentResource may be turned into an [InferredComponent] with [Component].
+type ComponentResource[I any, O pulumi.ComponentResource] interface {
+	// Construct a component resource
+	//
+	// ctx.RegisterResource needs to be called, but ctx.RegisterOutputs does not need to
+	// be called.
+	Construct(ctx *pulumi.Context, name, typ string, args I, opts pulumi.ResourceOption) (O, error)
+}
+
 // InferredComponent is a component resource inferred from code.
 //
-// To create an [InferredComponent], call the [Component] function.
+// To create an [InferredComponent], call the [ComponentF] function.
 type InferredComponent interface {
 	t.ComponentResource
 	schema.Resource
 
 	isInferredComponent()
 }
 
-func (derivedComponentController[R, I, O]) isInferredComponent() {}
+func (derivedComponentController[R, T, I, O]) isInferredComponent() {}
```

Comment:
Why are there now 4 type arguments? They used to represent the resource holder, the inputs and the component resource.

Reply:
True, and normally `derivedComponentController` extracts a type token plus annotations based on `R`.  However, in the special case of the functional style, it uses `O` because there is no `R`.  To support both modes, we need to abstract which type is reflected upon, hence `T`.

<img width="884" alt="image" src="https://github.com/user-attachments/assets/00ceb13a-8241-43ed-9958-227d62e6839d" />

<img width="876" alt="image" src="https://github.com/user-attachments/assets/0e8fad5d-1f27-43f5-bc4c-4d85f72968d5" />

Reply:
`Component` when took a function needed 2. `Component` when it took a struct needed 2. I don't understand why we would need to change it here. What do the 4 type arguments represent?

Reply:
I see three arguments, `Component[A,R,F]` before, and `ComponentF[A,O,F]` after.  R was renamed to O because R is typically the handler, not the component resource.

The fourth type argument is internal to `derivedComponentController` and simply serves to help it extract the type token from the appropriate source.